compiler/rustc_query_impl/src/lib.rs

```diff
@@ -17,37 +18,208 @@ extern crate rustc_macros;
 #[macro_use]
 extern crate rustc_middle;
 
-use rustc_data_structures::sync::AtomicU64;
+use memoffset::offset_of;
+use rustc_data_structures::fingerprint::Fingerprint;
+use rustc_data_structures::stable_hasher::HashStable;
 use rustc_middle::arena::Arena;
 use rustc_middle::dep_graph::{self, DepKindStruct};
 use rustc_middle::query::Key;
+use rustc_middle::ty::query::QueryCaches;
 use rustc_middle::ty::query::{
     query_keys, query_provided, query_provided_to_value, query_storage, query_values,
 };
-use rustc_middle::ty::query::{ExternProviders, Providers, QueryEngine};
+use rustc_middle::ty::query::{Providers, QueryEngine};
 use rustc_middle::ty::TyCtxt;
+use rustc_query_system::dep_graph::DepNodeParams;
+use rustc_query_system::ich::StableHashingContext;
 use rustc_span::Span;
+use std::marker::PhantomData;
 
 #[macro_use]
 mod plumbing;
-pub use plumbing::QueryCtxt;
+pub use plumbing::{Queries, QueryCtxt};
 use rustc_query_system::query::*;
 #[cfg(parallel_compiler)]
 pub use rustc_query_system::query::{deadlock, QueryContext};
 
 pub use rustc_query_system::query::QueryConfig;
+use rustc_query_system::HandleCycleError;
 
 mod on_disk_cache;
 pub use on_disk_cache::OnDiskCache;
 
 mod profiling_support;
 pub use self::profiling_support::alloc_self_profile_query_strings;
 
-rustc_query_append! { define_queries! }
+trait Bool {
+    fn value() -> bool;
+}
+
+struct True;
+
+impl Bool for True {
+    fn value() -> bool {
+        true
+    }
+}
+struct False;
+
+impl Bool for False {
+    fn value() -> bool {
+        false
+    }
+}
+
+#[derive(Copy, Clone)]
+struct DynamicQueryStatic {
+    name: &'static str,
+    query_state: usize,
+    query_cache: usize,
```

Comment:
There is nontrivial stuff going on with those fields. Could you doc-comment them?

Reply:
Does the `field-offset` crate make these clear enough?